src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,493 @@
+
+	RangeTableEntryDetails *rangeTableEntryDetails =
+		GetNextRTEToConvertToSubquery(query->jointree, conversionCandidates,
+									  plannerRestrictionContext,
+									  resultRelationId);
+
+	while (ShouldConvertLocalTableJoinsToSubqueries(query, rangeTableList,
+													resultRelationId,
+													plannerRestrictionContext))
+	{
+		ReplaceRTERelationWithRteSubquery(
+			rangeTableEntryDetails->rangeTableEntry,
+			rangeTableEntryDetails->restrictionList,
+			rangeTableEntryDetails->requiredAttributeNumbers,
+			context);
+		RemoveFromConversionCandidates(
+			conversionCandidates,
+			rangeTableEntryDetails->rangeTableEntry->relid);
+
+		rangeTableEntryDetails =
+			GetNextRTEToConvertToSubquery(query->jointree, conversionCandidates,
+										  plannerRestrictionContext,
+										  resultRelationId);
+	}
+}
+
+
+/*
+ * GetRangeTableEntriesFromJoinTree gets the range table entries that are
+ * on the given join tree.
+ */
+static void
+GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+								 List **joinRangeTableEntries)
+{
+	if (joinNode == NULL)
+	{
+		return;
+	}
+	else if (IsA(joinNode, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) joinNode;
+		Node *fromElement;
+
+		foreach_ptr(fromElement, fromExpr->fromlist)
+		{
+			GetRangeTableEntriesFromJoinTree(fromElement, rangeTableList,
+											 joinRangeTableEntries);
+		}
+	}
+	else if (IsA(joinNode, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) joinNode;
+		GetRangeTableEntriesFromJoinTree(joinExpr->larg, rangeTableList,
+										 joinRangeTableEntries);
+		GetRangeTableEntriesFromJoinTree(joinExpr->rarg, rangeTableList,
+										 joinRangeTableEntries);
+	}
+	else if (IsA(joinNode, RangeTblRef))
+	{
+		int rangeTableIndex = ((RangeTblRef *) joinNode)->rtindex;
+		RangeTblEntry *rte = rt_fetch(rangeTableIndex, rangeTableList);
+		*joinRangeTableEntries = lappend(*joinRangeTableEntries, rte);
+	}
+	else
+	{
+		pg_unreachable();
+	}
+}
+
+
+/*
+ * GetNextRTEToConvertToSubquery returns the range table entry
+ * which should be converted to a subquery. It considers the local join policy
+ * and result relation.
+ */
+static RangeTableEntryDetails *
+GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+							  ConversionCandidates *conversionCandidates,
+							  PlannerRestrictionContext *plannerRestrictionContext, Oid
+							  resultRelationId)
+{
+	RangeTableEntryDetails *localRTECandidate = NULL;
+	RangeTableEntryDetails *distributedRTECandidate = NULL;
+
+	if (list_length(conversionCandidates->localTableList) > 0)
+	{
+		localRTECandidate = linitial(conversionCandidates->localTableList);
+	}
+	if (list_length(conversionCandidates->distributedTableList) > 0)
+	{
+		distributedRTECandidate = linitial(conversionCandidates->distributedTableList);
+	}
+
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_LOCAL)
+	{
+		return localRTECandidate ? localRTECandidate : distributedRTECandidate;
```

Comment:
I think if `distributedRTECandidate ` is NULL than that mean we need to stop, so it should be okay to return `NULL` in that case.

I guess the above check wouldn't eliminate the checks because say that we have the following query:

`UPDATE local .... FROM dist` then the only candidate will be `dist` table since we cannot convert the result relation to a subqery. In that case `localRTECandidate` will be `NULL`, and `distributedRTECandidate` will refer to the `dist` table, even if the join policy is `prefer-local` we still should return the distributed table in this example